vispy/visuals/_scalable_textures.py

```diff
@@ -295,29 +298,31 @@ def clim_normalized(self):
 
     @staticmethod
     def _scale_data_on_cpu(data, clim, copy=True):
-        if copy:
-            should_cast_to_f32(data.dtype)
-            data = np.array(data, dtype=np.float32, copy=copy)
-        elif not copy and not np.issubdtype(data.dtype, np.floating):
-            raise ValueError("Data must be of floating type for no copying to occur.")
-
+        data = np.array(data, dtype=np.float32, copy=copy)
         if clim[0] != clim[1]:
+            # we always must copy the data if we change it here, otherwise it might change
+            # unexpectedly the data held outside of here
+            if not copy:
+                data = data.copy()
             data -= clim[0]
-            data *= 1.0 / (clim[1] - clim[0])
-        if should_cast_to_f32(data.dtype):
-            data = data.astype(np.float32)
+            data *= 1 / (clim[1] - clim[0])
         return data
 
     def scale_and_set_data(self, data, offset=None, copy=True):
         """Upload new data to the GPU, scaling if necessary."""
-        self._data_dtype = data.dtype
+        if self._data_dtype is None:
+            data.dtype == self._data_dtype
```

Comment:
```suggestion
        if self._data_dtype is None:
            self._data_dtype = data.dtype
```

@brisvag sorry for digging up this old code, but I got here looking through warnings that come up in tests. Should this be as above? Otherwise I'm not sure what this line is meant to be doing.

Reply:
oof... It's been some time, not sure I remember much about these changes. Your change definitely looks like the most likely intended code ^^' That probably means we have some test missing that should catch a problem here.